app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.library.bookmarks
+
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarkNodeType
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotSame
+import org.junit.Test
+
+class BookmarkStoreTest {
+
+    @Test
+    fun `change the tree of bookmarks starting from an empty tree`() = runBlocking {
+        val initialState = BookmarkState(null)
+        val store = BookmarkStore(initialState)
+
+        store.dispatch(BookmarkAction.Change(tree)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(tree, store.state.tree)
+        assertEquals(BookmarkState.Mode.Normal, store.state.mode)
+    }
+
+    @Test
+    fun `select and deselect bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree)
+        val store = BookmarkStore(initialState)
+
+        store.dispatch(BookmarkAction.Select(childItem)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(BookmarkState.Mode.Selecting(setOf(childItem)), store.state.mode)
+
+        store.dispatch(BookmarkAction.Deselect(childItem)).join()
+
+        assertEquals(BookmarkState.Mode.Normal, store.state.mode)
+    }
+
+    @Test
+    fun `deselect all bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree, BookmarkState.Mode.Selecting(setOf(item, childItem)))
+        val store = BookmarkStore(initialState)
+
+        assertEquals(BookmarkState.Mode.Selecting(setOf(item, childItem)), store.state.mode)
+
+        store.dispatch(BookmarkAction.DeselectAll).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(BookmarkState.Mode.Normal, store.state.mode)
+    }
+
+    @Test
+    fun `deleting bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree, BookmarkState.Mode.Selecting(setOf(item, childItem)))
+        val store = BookmarkStore(initialState)
+
+        assertEquals(BookmarkState.Mode.Selecting(setOf(item, childItem)), store.state.mode)
+
+        val newTree = BookmarkNode(
+            BookmarkNodeType.FOLDER,
+            "123",
+            null,
+            0,
+            "Mobile",
+            null,
+            listOf(separator, subfolder)
+        )
+
+        store.dispatch(BookmarkAction.Change(newTree)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(newTree, store.state.tree)
+        assertEquals(BookmarkState.Mode.Normal, store.state.mode)
```

Comment:
Missing test: new incoming tree when currently selected items appear in the new tree. You should end up in the `selecting` mode.

Reply:
Done.